src/altimeter.rs

```diff
@@ -0,0 +1,273 @@
+use bmp388::{config::{FifoConfig, OversamplingConfig}, Blocking, PowerControl, BMP388};
+use heapless::Vec;
+use core::{cell::RefCell, ptr::addr_of_mut, sync::atomic::AtomicBool};
+use cortex_m::interrupt::Mutex;
+use stm32f4xx_hal::{
+    i2c::dma::{
+            I2CMasterHandleIT, I2CMasterWriteReadDMA, I2cCompleteCallback
+        }, interrupt
+};
+use stm32f4xx_hal::i2c;
+use stm32f4xx_hal::timer::delay::SysDelay;
+use crate::futures::YieldFuture;
+use crate::pins::Altimeter;
+use crate::bmp581::I2c1Handle;
+use crate::bmp581::BMP581;
+
+const BMP388_FRAME_COUNT: usize = 73;
+const BMP388_BUF_SIZE:    usize = 512;
+const BMP581_FRAME_COUNT: usize = 16;
+const BMP581_BUF_SIZE:    usize = BMP581_FRAME_COUNT * 6;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Default)]
+pub struct PressureTemp {
+    pub pressure: u32,
+    pub temperature: u32,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct FifoFrames([PressureTemp; ALTIMETER_FRAME_COUNT]);
+
+impl Default for FifoFrames {
+    fn default() -> Self {
+        FifoFrames([PressureTemp::default(); BMP388_FRAME_COUNT])
+    }
+}
+
+impl From<[PressureTemp; ALTIMETER_FRAME_COUNT]> for FifoFrames {
+    fn from(frames: [PressureTemp; ALTIMETER_FRAME_COUNT]) -> Self {
+        FifoFrames(frames)
+    }
+}
+
+impl Into<[PressureTemp; ALTIMETER_FRAME_COUNT]> for FifoFrames {
+    fn into(self) -> [PressureTemp; ALTIMETER_FRAME_COUNT] {
+        self.0
+    }
+}
+
+impl<'a> Into<&'a [PressureTemp; ALTIMETER_FRAME_COUNT]> for &'a FifoFrames {
+    fn into(self) -> &'a [PressureTemp; ALTIMETER_FRAME_COUNT] {
+        &self.0
+    }
+}
+
+pub struct BMP388Wrapper {
+    pub bmp: BMP388<I2c1Handle, Blocking>
+}
+
+impl BMP388Wrapper {
+    pub fn new(i2c: I2c1Handle, delay: &mut SysDelay) -> Self {
+        let mut bmp = BMP388::new_blocking(i2c, Self::ADDRESS, delay).unwrap();
+        bmp.set_fifo_config(FifoConfig {
+            enabled: true,
+            stop_on_full: false,
+            store_pressure: true,
+            store_temperature: true,
+            return_sensor_time: false,
+            subsampling: 0,
+            filter_data: false
+        }).unwrap();
+        Self { bmp }
+    }
+}
+
+pub trait AltimeterFifoDMA<
+    const FRAMES: usize, const BUF_SIZE: usize
+>: I2CMasterWriteReadDMA {
+    const FIFO_READ_REG: u8;
+    const ADDRESS: u8;
+    fn dma_interrupt(&mut self);
+    fn process_fifo_buffer(data: [u8; BUF_SIZE]) -> [PressureTemp; FRAMES];
+}
+
+fn frame_to_reading(pres: &[u8], temp: &[u8]) -> PressureTemp {
+    PressureTemp {
+        pressure:    (pres[0] as u32) | (pres[1] as u32) << 8 | (pres[2] as u32) << 16,
+        temperature: (temp[0] as u32) | (temp[1] as u32) << 8 | (temp[2] as u32) << 16,
+    }
+}
```

Comment:
for some reason the temperature compensation is very necessary for the 388, otherwise reports temps like 130C, could we modify PressureTemp to hold f32's and then use the `BMP388::compensated_pressure` (& temperature) methods. I think we should also implement these for the 581.